### PR TITLE
Assert layer ordering in `layers_that_should_push_properties`

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR adds two new assertion:

* `TreeSynchronizer::PushLayerProperties` that verifies that the source layers are iterated in a stable order. 
* `Layer::Layer`: to assert that the layer ids are generated in the same order.

It further changes the assertion in `PictureLayer::DropRecordingSourceContentIfInvalid` to use the `Layer->id()` instead of the pointer to avoid having to track two identifiers for `Layer` that must be stable (`Layer::id` is used to sort the layers in `TreeSynchronizer`). 

Part of https://linear.app/replay/issue/RUN-1229
[In detail analysis](https://linear.app/replay/issue/RUN-1229/mismatch-in-ccpicturelayerdroprecordingsourcecontentifinvalid#comment-0b38e10e)

### CODE REVIEW

File: cc/layers/layer.cc [hunk 1]
```diff
@@ -119,14 +129,9 @@ Layer::Layer()
       property_tree_sequence_number_(-1),
       ignore_set_needs_commit_for_test_(false),
       bitflags_(0u),
-      subtree_property_changed_(false) {
-  // https://linear.app/replay/issue/RUN-885
-  recordreplay::RegisterPointer("Layer", this);
```
Comment (kannanvijayan) [inline cc/layers/layer.cc:124]: I'm less sure about removing the existing registration here.  There may be other places / asserts where we use `GetPointerId()` to map the raw pointer to a stable id.  But other than that this looks ok.

File: cc/layers/layer.cc [hunk 2]
```diff
@@ -73,6 +73,16 @@ static_assert(sizeof(Layer) == sizeof(SameSizeAsLayer),
               "Layer should stay small");
 
 base::AtomicSequenceNumber g_next_layer_id;
+size_t next_layer_lock = recordreplay::CreateOrderedLock("Layer::NextID");
+
+// https://linear.app/replay/issue/RUN-1229
+inline int NextLayerId() {
+  recordreplay::OrderedLock(next_layer_lock);
```
Comment (kannanvijayan) [inline cc/layers/layer.cc:80]: This is a good find an honestly a likely root cause of the later mismatch.  @bhackett1024 thoughts?

**Review by kannanvijayan — COMMENTED** [review]
Looks good.  I'm waiting for @bhackett1024 's commentary on some of the trickier questions about whether the ordering of the atomic incremeneting id-mechanism is a possible root cause.

File: cc/layers/layer.cc (hunk 1 above)
Comment (bhackett1024) [reply cc/layers/layer.cc:124]: We don't really have a developed strategy to make sure we keep track of all the places where PointerId() is used on different types and makes it easy to reason about removing unnecessary RegisterPointer calls.  There is a convention of adding comments pointing to the usage or issues which the RegisterPointer is needed for, but that isn't always going to be present or guaranteed to be up to date.

If there are still uses of PointerId(), they could lead either to bogus mismatches that test the pointer ID, or crashes when CompareByPointerId or a similar class is used to sort objects by their pointer ID.  The former case won't cause crashes for users but will show up in crash reports and can be fixed by there, the latter case should be caught by the testing we do in CI before releasing a build.

In this case I think it should be fine to remove this RegisterPointer call.

File: cc/layers/layer.cc (hunk 2 above)
Comment (bhackett1024) [reply cc/layers/layer.cc:80]: It's possible, generally I'd prefer it though if we made sure a thread ordering issue was definitely a problem before adding more ordered locks.  It's not clear from the chromium commit whether layers can be constructed on multiple threads in a normal render process.  In this case we could add an assertion in the layer constructor on its ID and see if that gets tripped.  This isn't a big deal though and it's fine to land the patch as is.

**Review by bhackett1024 — APPROVED** [review]

File: cc/trees/tree_synchronizer.cc [hunk 3]
```diff
@@ -219,6 +219,11 @@ void TreeSynchronizer::PushLayerProperties(
       commit_state.layers_that_should_push_properties.end();
   for (auto it = source_layers_begin; it != source_layers_end; ++it) {
     auto* source_layer = *it;
+    // https://linear.app/replay/issue/RUN-1229
+    recordreplay::Assert(
```
Comment (Domiii) [inline cc/trees/tree_synchronizer.cc:223]: hi @MichaReiser,
This is awesome stuff!

When I last looked, I did not pay enough attention, and forgot to mention: When adding new or revisiting old `Assert`s, we want to start adding a more specific label. We use these labels to do some automagical crash filtering in http://admin/crash/filters

Currently the label format is `[RUN-X]`, where `X` refers to the original crash report issue.

In the future, we want to move toward `[RUN-X-Y]`, where `Y` is the child issue that the PR is filed under; hopefully by next week. It is currently blocked by https://linear.app/replay/issue/SCS-784/[crash-buckets]-allow-more-flexible-labels

**Pull request merged**